[core] Use datadog_id label for service discovery.

geowa4 opened the pull request:
This still falls back to parsing the image name when this label is not
present.

I created this as a result of #2576 and my conversation with support.

geowa4 commented:
The unit test failure appears to be in installing Kong and unrelated to the pull request.

hkaj commented:
Hi @geowa4, thanks for submitting a PR. I understand your issue and it's definitely worth fixing, but I'm not sure this is the right approach. We should probably just update the path we query in the kv store to include repo and tag.

The problem with using something else is that we're already watching the config store for changes in keys and values to trigger a configuration reload in the agent. Moving this information somewhere else would require adding more calls to external APIs for this purpose and we don't want that.

Would it work for you if we move to `repository/path/image_name:image_tag` in the store instead of `image_name` only? Possibly replacing `/` with `-`, I didn't try it yet.

geowa4 commented:
Hi @hkaj, I'm not sure I understand your opposition. This requires no additional calls to any other APIs.

hkaj commented:
@geowa4 sorry that wasn't clear.

What I'm thinking about is the case where you want to add or remove a check for a container. In this case you would modify the label to add/remove a key where the config template can be found.
That means the agent has to inspect containers to see if labels were added, removed or modified since its previous run and then query the new path or stop querying an old one.

That's more expensive to perform than just watching the canonical path (`/datadog/check_configs` by default). Especially because we already watch this path for changes in configuration templates (added tags, modified options, etc.).

Improving the path achieves the same goal without needing to inspect containers for updated labels.

geowa4 commented:
@hkaj, I think I'm still missing something. All I've done with this change is offer an alternative to parsing the image name for a magic string. From what I gather, you take that parsed string and append it to a predefined path like so: `/datadog/check_configs/$PARSED_IMAGE_NAME`. What I have submitted in this PR is optionally using an explicit label to replace that parsing, making the path `/datadog/check_configs/$SPECIAL_LABEL_VALUE`.

There is no continuous checking for updates to any labels. The mode of operation is exactly the same as it is now, which is how you want it. Also, I don't think you can change the labels of a running container.

hkaj commented:
ooh my bad, I didn't get it correctly. It shouldn't be too expensive to run actually.

Let me give you some more context:

Right now at each run the agent looks for changes in the config store[0][1] and triggers a configuration reload[2] if changes happened (templates were edited, added, removed, or checks were added/removed for some containers).
So let's say you have an nginx container with the nginx check enabled through a config template. If you want to also enable the http_check against it you update the template and the agent will pick-up the new config without having to restart.

Now if we use container labels to specify which template to query and apply to each container we'll also need to look for changes in labels for every container.

That's not as expensive as I thought since we can simply keep a map of these labels somewhere and run `docker.Client().containers(filter=blah)` to get the current one at each run and update it, but fixing the path issue by using the full image name (repo and tag included) makes this part unnecessary and thus seems more efficient to me.

Let me know what you think.

[0]: https://github.com/DataDog/dd-agent/blob/55feac7d7549361e929f5586e05c8bd4ea591802/agent.py#L201-L210
[1]: https://github.com/DataDog/dd-agent/blob/25b130b05ce70f79353005515bb3cd29a4a0be93/utils/service_discovery/abstract_config_store.py#L155-L175
[2]: https://github.com/DataDog/dd-agent/blob/55feac7d7549361e929f5586e05c8bd4ea591802/agent.py#L212-L219

geowa4 commented:
> Now if we use container labels to specify which template to query and apply to each container we'll also need to look for changes in labels for every container.

I don't understand why that's true. Each container has many labels just as it has exactly one image name. For a running container, labels change as often as images names, which is never. This PR introduces a special label so we can amend that statement to "each container has exactly one `datadog_id` label just as it has exactly one image name". Logically, these are now equivalent.

I do not support replacing characters in the image name. This practice still loses information provided by the user; it just loses less than the current implementation. While it is an improvement, it is not sufficient, especially when other more explicit solutions exist. For example, if someone is refactoring in a way requiring different checks and changes her images from `private-registry.com/foo/bar:web` to `private-registry.com/foo/bar-web`, you're going to make her angry. This is not acceptable.

geowa4 commented:
The effect of my proposed change:
```
$ python
Python 2.7.11 (default, Dec  5 2015, 20:44:19)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from utils.dockerutil import DockerUtil
>>> docker_client = DockerUtil().client
>>> docker_client.containers()
[{u'Status': u'Up 45 hours', u'Created': 1465399125, u'Image': u'nginx', u'Labels': {u'datadog_id': u'custom-nginx'}, u'NetworkSettings': {u'Networks': {u'bridge': {u'NetworkID': u'', u'MacAddress': u'', u'GlobalIPv6PrefixLen': 0, u'Links': None, u'GlobalIPv6Address': u'', u'IPv6Gateway': u'', u'IPAMConfig': None, u'EndpointID': u'', u'IPPrefixLen': 16, u'IPAddress': u'172.17.0.2', u'Gateway': u'172.17.0.1', u'Aliases': None}}}, u'HostConfig': {u'NetworkMode': u'default'}, u'ImageID': u'sha256:6f8d099c3adc0f9c14bdc11d6c698a647c2f0457aa1ae76a592d6f7dc9bdb5b8', u'State': u'running', u'Command': u"nginx -g 'daemon off;'", u'Names': [u'/stoic_allen'], u'Mounts': [], u'Id': u'445272bc04e4ad5f0d0a480e7a56af36aeef57053601835ef3579d98cdf3e567', u'Ports': [{u'IP': u'0.0.0.0', u'Type': u'tcp', u'PublicPort': 32770, u'PrivatePort': 443}, {u'IP': u'0.0.0.0', u'Type': u'tcp', u'PublicPort': 32771, u'PrivatePort': 80}]}]
>>> docker_client.containers()[0].get('Image').split(':')[0].split('/')[-1]
u'nginx'
>>> docker_client.containers()[0].get('Labels', {}).get('datadog_id')
u'custom-nginx'
```

The `container` dictionary also has a `Labels` key that is a dictionary of all labels applied to the container. My change simply looks for a key `datadog_id` in that dictionary. If it does not exist, it falls back to the equivalent of the original code which uses the `Image` key.

danroyo commented on utils/service_discovery/sd_docker_backend.py:
```diff
@@ -131,11 +131,17 @@ def _get_kube_config(self, c_id, key):
                     return pod.get(key, {})
 
+    def _get_config_id(self, container):
+        config_id = container.get('Labels', {}).get('datadog_id')
```
If the first `get` call uses the fallback value, the second call here is going to throw a KeyError

geowa4 replied:
It doesn't, but I updated the code to be very explicit and avoid confusion.

From my REPL:
```
>>> {}.get('datadog_id') is None
True
```

geowa4 commented:
Superseded by #2593.
